ocaml/xapi/xapi_pool.ml

```diff
@@ -781,15 +781,39 @@ let eject ~__context ~host =
 
 	if Pool_role.is_master () then raise Cannot_eject_master
 	else begin
-		(* Fail the operation if any VMs are running here (except control domains) *)
+		(* Fail the operation if any VMs are running here (except dom0) *)
 		let my_vms_with_records = Db.VM.get_records_where ~__context ~expr:(Eq(Field "resident_on", Literal (Ref.string_of host))) in
 		List.iter (fun (_, x) -> 
-			if (not x.API.vM_is_control_domain) && x.API.vM_power_state<>`Halted
+			if (not (Helpers.is_domain_zero ~__context (Db.VM.get_by_uuid ~__context ~uuid:x.API.vM_uuid))) && x.API.vM_power_state <>`Halted
 			then begin
 				error "VM uuid %s not in Halted state and resident_on this host" (x.API.vM_uuid);
 				raise (Api_errors.Server_error(Api_errors.operation_not_allowed, ["VM resident on host"]))
 			end) my_vms_with_records;
 
+		(* all control domains resident on me should be destroyed once I leave the
+		pool, therefore pick them out as follows: if they have a valid resident_on,
+		the latter should be me; if they don't (e.g. they are halted), they should have
+		disks on my local storage *)
+		let control_domains_to_destroy = List.filter (fun (vm_ref, vm_rec) ->
+			if not (vm_rec.API.vM_is_control_domain) then
+				false
+			else if Db.is_valid_ref __context vm_rec.API.vM_resident_on then
+				vm_rec.API.vM_resident_on = host
+			else
+				let resident_host =
+				(try
+					let filter_valid x = List.filter (fun y -> Db.is_valid_ref __context y) x in
+					let vbds = Db.VM.get_VBDs ~__context ~self:vm_ref in
+					let vdis = List.map (fun v -> Db.VBD.get_VDI ~__context ~self:v) (filter_valid vbds) in
+					let srs = List.map (fun v -> Db.VDI.get_SR ~__context ~self:v) (filter_valid vdis) in
+					let local_srs = List.filter (fun x -> not (Helpers.is_sr_shared ~__context ~self:x)) (filter_valid srs) in
+					let sr = List.hd (filter_valid local_srs) in
+					let pbds = Db.SR.get_PBDs ~__context ~self:sr in
+					let hosts = List.map (fun p -> Db.PBD.get_host ~__context ~self:p) (filter_valid pbds) in
+					List.hd (filter_valid hosts)
+				with _ -> Ref.null) in
```

Comment:
I've changed the code according to Rob's suggestion, so we don't need that block any more.